extra_geom/detectors.py

```diff
@@ -854,6 +854,276 @@ def to_distortion_array(self, allow_negative_xy=False):
+    #       (lengths 8, 512, 128). ss/fs are slow-scan and fast-scan.
+
+    #     Returns
+    #     -------
+    #     out : ndarray
+    #       Array with the one dimension fewer than the input.
+    #       The last two dimensions represent pixel y and x in the detector space.
+    #     centre : ndarray
+    #       (y, x) pixel location of the detector centre in this geometry.
+    #     """
+    #     assert data.shape == (8, 512, 128)
+    #     size_yx, centre = self._get_dimensions()
+    #     tmp = np.empty((8 * 8,) + size_yx, dtype=data.dtype)
+
+    #     for i, (module, mod_data) in enumerate(zip(self.modules, data)):
+    #         tiles_data = np.split(mod_data, 8)
+    #         for j, (tile, tile_data) in enumerate(zip(module, tiles_data)):
+    #             # We store (x, y, z), but numpy indexing, and hence affine_transform,
+    #             # work like [y, x]. Rearrange the numbers:
+    #             fs_vec_yx = tile.fs_vec[:2][::-1]
+    #             ss_vec_yx = tile.ss_vec[:2][::-1]
+
+    #             # Offset by centre to make all coordinates positive
+    #             corner_pos_yx = tile.corner_pos[:2][::-1] + centre
+
+    #             # Make the rotation matrix
+    #             rotn = np.stack((ss_vec_yx, fs_vec_yx), axis=-1)
+
+    #             # affine_transform takes a mapping from *output* to *input*.
+    #             # So we reverse the forward transformation.
+    #             transform = np.linalg.inv(rotn)
+    #             offset = np.dot(rotn, corner_pos_yx)  # this seems to work, but is it right?
+
+    #             affine_transform(
+    #                 tile_data,
+    #                 transform,
+    #                 offset=offset,
+    #                 cval=np.nan,
+    #                 output_shape=size_yx,
+    #                 output=tmp[i * 8 + j],
+    #             )
+
+    #     # Silence warnings about nans - we expect gaps in the result
+    #     with warnings.catch_warnings():
+    #         warnings.simplefilter("ignore", category=RuntimeWarning)
+    #         out = np.nanmax(tmp, axis=0)
+
+    #     return out, centre
+
+    def _get_dimensions(self):
+        """Calculate appropriate array dimensions for assembling data.
+
+        Returns (size_y, size_x), (centre_y, centre_x)
+        """
+        corners = []
+        for module in self.modules:
+            for tile in module:
+                corners.append(tile.corners())
+        corners = np.concatenate(corners)[:, :2] / self._pixel_shape
+
+        # Find extremes, add 1 px margin to allow for rounding errors
+        min_xy = corners.min(axis=0).astype(int) - 1
+        max_xy = corners.max(axis=0).astype(int) + 1
+
+        size = max_xy - min_xy
+        centre = -min_xy
+        # Switch xy -> yx
+        return tuple(size[::-1]), centre[::-1]
+
+    @staticmethod
+    def split_tiles(module_data):
+        # 2 rows of 8 ASICs each. This slicing is faster than np.split().
```

Comment:
Perhaps this is the point to discuss: do we also need two rows for AGIPD-1M? I initially copied a CrystFEL geometry in defining a module as a single row of 8 panels, 128 x 64 each. I've also seen in the calibration code that it's actually two rows of 64 x 64 ASICs. If there's no gap between the rows, maybe it doesn't matter for the geometry? But we should make sure it's clearly documented in either case.

Reply:
yep, That's probably fine to consider the 2 asics as 1 tile as long as there is no gap and the data is aligned.

Reply:
Have we got some confirmation that that's how they are?

Reply:
> The dead area of the final system is minimized for each FEM by using a monolithic sensor bump-bonded to 2 × 8 individual ASICs. The large sensor guarantees that there are no blind spots within the sensitive surface (0% dead area); however, since the ASICs cannot physically touch each other, the pixels horizontally in between two ASICs are twice as wide (400 µm × 200 µm), covering the area of two ordinary pixels. Pixels vertically in between two ASICs are of normal size.

https://journals.iucr.org/s/issues/2019/01/00/gb5080/

Reply:
Also (even if that's not a formal proof), there are no gap visible between asics looking at rings

![asic](https://user-images.githubusercontent.com/32831491/93761100-7cfdbe80-fc0d-11ea-9bdf-a7ded92c3c94.png)

Reply:
Also confirmed here: https://xfel.tind.io/record/395/ (chapter 5)

Reply:
OK, sounds good.

Reply:
& thanks for looking into this :-)